compiler/src/dotty/tools/dotc/core/Symbols.scala

```diff
@@ -558,32 +558,44 @@ object Symbols {
     /** If this is a top-level class, and if `-Yretain-trees` is set, return the TypeDef tree
      *  for this class, otherwise EmptyTree. This will force the info of the class.
      */
-    def tree(implicit ctx: Context): tpd.Tree /* tpd.TypeDef | tpd.EmptyTree */ = {
+    def tree(implicit ctx: Context): tpd.Tree /* tpd.TypeDef | tpd.EmptyTree */  = {
+      import ast.Trees._
+      def findTree(tree: tpd.Tree): Option[tpd.TypeDef] = tree match {
+        case PackageDef(_, stats) =>
+          stats.flatMap(findTree).headOption
+        case tree: tpd.TypeDef if tree.symbol == this =>
+          Some(tree)
+        case _ =>
+          None
+      }
+      val t = unitTree
+      if (t.isEmpty) t
+      else findTree(tree).get
+    }
+
+    /** If this is a top-level class, and if `-Yretain-trees` or `-YlinkOptimise` is set,
+      * return the PackageDef tree for this class, otherwise EmptyTree.
+      * This will force the info of the class.
+      */
+    def unitTree(implicit ctx: Context): tpd.Tree /* tpd.PackageDef | tpd.TypeDef | tpd.EmptyTree */ = {
```

Comment:
I'm unhappy about having both `unitTree` and `tree`. I think we should just have `unitTree` (but give it a different name like just `tree` because it's not the full compilation unit tree, it's only the current class tree wrapped with PackageDefs).